pkg/tenantctl/controller.go

```diff
@@ -157,219 +172,470 @@ func (c *Controller) processNextWorkItem() bool {
 	if err != nil {
-		err = errors.WithMessagef(err, "Failed to update tenant '%s' in namespace '%s'", tenant.GetName(), tenant.GetNamespace())
-		log.Printf("ERROR: %s", err.Error())
-		return nil, err
+		err = errors.WithMessagef(err,
+			"failed to update tenant %q in namespace %q",
+			tenant.GetName(), tenant.GetNamespace(),
+		)
+		c.logPrintln(tenant, err)
+		return tenant, err
 	}
-	return updatedTenant, nil
+	return result, nil
 }
 
-// An error is returned in cases to signalize processNextWorkItem() to retry processing the tenant.
-// If no error is returned this signalized OK, do not retry. This should be done in cases where retry will not help.
-func (c *Controller) handleError(tenant *api.Tenant, err error, result api.TenantResult) error {
-	log.Printf("ERROR: %s", err.Error())
-	tenant.Status.Result = result
-	tenant.Status.Message = utils.Trim(err.Error())
-	_, updateStatusErr := c.updateStatus(tenant)
-	return updateStatusErr
+func (c *Controller) checkNamespaceExists(name string) (bool, error) {
+	namespaces := c.factory.CoreV1().Namespaces()
+	namespace, err := namespaces.Get(name, metav1.GetOptions{})
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return false, nil
+		}
+		err = errors.WithMessagef(err, "error: failed to get namespace %q", name)
+		return false, err
+	}
+	return namespace.GetDeletionTimestamp().IsZero(), nil
 }
 
-func (c *Controller) rollbackIfRequired(tenantKey string) {
-	tenant, err := c.fetcher.ByKey(tenantKey)
+func (c *Controller) createTenantNamespace(config clientConfig, tenant *api.Tenant) (string, error) {
+	c.logPrintln(tenant, "creating new tenant namespace")
+	namespaceManager := c.getNamespaceManager(config)
+	nsName, err := namespaceManager.Create(tenant.GetName(), nil)
 	if err != nil {
-		log.Printf("ERROR: Could not get tenant during rollback: %s", err.Error())
+		err = errors.WithMessage(err, "failed to create new tenant namespace")
+		c.logPrintln(tenant, err)
+		return "", err
 	}
-	if tenant.Status.Progress != api.TenantProgressFinished {
-		_ = c.rollback(tenant)
+	return nsName, err
+}
+
+func (c *Controller) rollbackTenantNamespace(namespace string, tenant *api.Tenant, config clientConfig) error {
+	if namespace == "" {
+		return nil
 	}
+	c.logPrintf(tenant, "rolling back tenant namespace %q", namespace)
+	namespaceManager := c.getNamespaceManager(config)
+	err := namespaceManager.Delete(namespace)
+	if err != nil {
+		err = errors.WithMessagef(err, "failed to rollback tenant namespace %q", namespace)
+		c.logPrintln(tenant, err)
+		return err
+	}
+	return nil
 }
 
-func (c *Controller) rollback(tenant *api.Tenant) error {
-	log.Printf("Rollback tenant %s", tenant.GetName())
-	if tenant.Status.TenantNamespaceName == "" {
-		log.Printf("Nothing to rollback for tenant %s", tenant.GetName())
-	} else {
-		err := c.deleteNamespace(tenant)
+func (c *Controller) syncTenantRoleBinding(tenant *api.Tenant, namespace string, config clientConfig) (bool, error) {
```

Comment:
Not sure if I like how the `(syncNeeded bool, err error)` return parameters are used here. Typically one would expect either `syncNeeded` is set to a meaningful value **or** an error occured. But here it is used both in mixed combinations. It could be even inconsistent, e.g. if the error in line https://github.com/SAP/stewardci-core/pull/14/files#diff-68316b8747cd48321fb190be91cc7d5fR469 occurs - we don't know if sync is needed.
It is also not clear, what `syncNeeded` means. "sync has been needed" if error is null and "sync is still needed" if error is set?
API documentation would also help here.

Reply:
Tackled in 72e8a0b14d98b755f66e83c3ee9748ddca7368f1.

I've clarified the meaning in the methods documentation but kept the output parameter as such (renamed to `updateNeeded` to avoid the term `sync`).